Policy and Deprecation into for README.md (#26)
drowe67/codec2: README.md

  + APIs for packet data over radio [README_data](README_data.md)
  + An STM32 embedded version of FreeDV 1600/700D/700E for the [SM1000](stm32/README.md)

## Old Code and Deprecated Features

In July 2023 this repo was refactored, older code can be found in https://github.com/drowe67/codec2-dev. 

We are currently conducting a major re-development of Codec 2, new speech coding, modems, and FreeDV modes are under active development. As such, we have limited resources available for mantinence or development on deprecated features unless there is a major bug. We'd rather put our efforts into new features! Deprecated modes include:

> 1. FreeDV 1600, 700C, 700D, 700E, 2020

Reviewer: Generally I would not deprecate modes until there's something available to replace them. It would probably be more accurate just to say that we're only accepting bug fixes for the stuff we have now. If we are deprecating stuff, I'd recommend deprecating the following:

* 700C
* 2020/2020B
* Codec2 modes *besides* 1600/3200 (for M17) and whatever else is used by 700D/E/1600.
* LPCNet

Author: I agree with:
> Generally I would not deprecate modes until there's something available to 
> replace them. 

But don't feel this captures our PLT decision:
> It would probably be more accurate just to say that we're only accepting bug 
> fixes for the stuff we have now.
(unless it is a major bug - which I have covered in the text above.)

Just because we have it now, is not a good reason for our resources to be diverted away from important R&D to work on it.

The key point here is to avoid diverting important resources to non-critical bug fixes and maintenance to code that is hopefully (if we meet our performance goals) going to be deprecated soon.

Author: @tmiw - I've had another go at the wording in https://github.com/drowe67/codec2/pull/26/commits/aa1d6c0b977a178de55c18a4b026d3c057f7d868 - see what you think

  + APIs for packet data over radio [README_data](README_data.md)
  + An STM32 embedded version of FreeDV 1600/700D/700E for the [SM1000](stm32/README.md)

## Old Code and Deprecated Features

In July 2023 this repo was refactored, older code can be found in https://github.com/drowe67/codec2-dev. 

We are currently conducting a major re-development of Codec 2, new speech coding, modems, and FreeDV modes are under active development.  We have limited resources available for maintenance or development on features that may soon be replaced (with the exception of major bugs).  We'd rather put our efforts into new features! Modes and Features we are not actively maintaining at present, and likely to be superseded in the near future include:
1. FreeDV 1600, 700C, 700D, 700E, 2020.

> 1. Codec 2 modes with the exception of Codec 2 3200 for M17.

Reviewer: Per page 28 of https://spec.m17project.org/, Codec2 1600 is also used.

```suggestion
1. Codec 2 modes with the exception of Codec 2 1600 and 3200 for M17.
```

Author: Woj has specifically asked us to improve 1600, so would deprecating/not maintaining the current 1600 be one of our aims?

Reviewer: Fair enough, we can keep as is.

1. Our fork of LPCNet.

> 1. The stm32/SM1000 firmware (paused until new modes available).

Reviewer: SM1000 is probably not a mode or feature in itself.

```suggestion

Development of the STM32/SM1000 firmware is also paused until new FreeDV modes become available.
```

Author: It's the same point as above that we want to capture with this section - we don't want to be diverting resources into SM1000 firmware maintenance right now.  When we have something new and exciting (that's working will on say x86), we'll schedule an upgrade.

Reviewer: I was mainly splitting out this into a separate category since the SM1000 not a Codec2 thing per se, just a client of it that also happens to require a significant number of changes to reduce requirements enough for it to work. :) Feel free to word as required.

The code supporting these modes won't be going away any time soon (and we will continue to include any modes/code in popular use), but we have chosen not to actively develop it at this time.

## Pull Requests, Feature Requests

We have a process for considering Feature Requests and Pull Requests that we will guide you through.

Feature Requests can be submitted via GitHub Issues.

Before writing any code or submitting a PR - **please discuss** the PR with developers by raising a GitHub Issue. We have many years of experience and a carefully considered plan for Codec 2 development, and can guide you on work that will most benefit this project.

## Ports to non gcc/C99 Compilers

> We have standardized on gcc/C99.  We and encourage people who want to use non-standard compilers like MSVC and certain embedded compilers to maintain their own Codec 2 forks. Our focus needs to be on what’s unique about our project – the speech codec and modem waveforms – rather than consuming time and resources on non-core activities that others can do equally well. If you decide to fork Codec 2 to a non C99/gcc compiler - please ensure you port the ctests and that they all pass.  If the tests have not been ported or do not pass - it's not Codec 2.

Reviewer: Should Clang get mentioned as well, or does that fall under gcc?

Author: Not sure ...  do we support/do any compiling with clang?  We should probably make it specific, so we know what we should/shouldn't support.  We have agreed to use clang for code formatting.

Reviewer: Clang is used for macOS and Windows binaries. Not sure what Linux distros use it by default if any.

Author: Cool. So what we're trying to do here is define what we support (and what we don't) so we can make calls on what we work on (and what we don't work on) should an issue arise.

I guess from our experience we can say that Clang works just as well as gcc, so there must be a very high level of compatibility (and C99 compliance), certainly no non-standard quirks like say MSVC.

If a clang build through up a bug, I would be tempted to look into it - cross compiling the same code is great way to find bugs in C.

Would it perhaps be better if we just say "C99"? Or should we add clang to the list?  Hmm, there might also be other compilers that work :thinking: So I'm tempted to say "C99, and we use gcc for development and testing".

Reviewer: Yeah, I think just mentioning C99 will work. 

(BTW on the freedv-gui side, I want to say that C++14 is mandatory but I'm not 100% sure. Regardless, Clang and GCC both compile it fine.)

   make
   ```

## Bulding for Windows

> We target our C code to compile with `gcc` to the [C99 standard](#ports-to-non-gccc99-compilers). We recommend using Linux to cross compile for Windows.

Reviewer: ```suggestion
We target our C code to compile with `gcc` to the [C99 standard](#ports-to-non-gccc99-compilers). We recommend using MinGW to cross compile for Windows.
```

(Note: freedv-gui uses LLVM MinGW to compile itself as well as Codec2/LPCNet.)
